client/client-lib/src/api.rs

```diff
@@ -198,3 +275,234 @@ impl<C: ClientT> WsFederationApi<C> {
+        ) -> std::result::Result<R, JsonRpcError>
+        where
+            R: jsonrpsee_core::DeserializeOwned,
+        {
+            let json = self.0.request(method).await?;
+            Ok(serde_json::from_str(&json).unwrap())
+        }
+
+        async fn notification<'a>(
+            &self,
+            _method: &'a str,
+            _params: Option<jsonrpsee_types::ParamsSer<'a>>,
+        ) -> std::result::Result<(), JsonRpcError> {
+            unimplemented!()
+        }
+
+        async fn batch_request<'a, R>(
+            &self,
+            _batch: Vec<(&'a str, Option<jsonrpsee_types::ParamsSer<'a>>)>,
+        ) -> std::result::Result<Vec<R>, JsonRpcError>
+        where
+            R: jsonrpsee_core::DeserializeOwned + Default + Clone,
+        {
+            unimplemented!()
+        }
+    }
+
+    fn federation_member<C: SimpleClient + Send + Sync>() -> FederationMember<Client<C>> {
+        FederationMember {
+            url: String::new(),
+            peer_id: PeerId::from(0),
+            client: RwLock::new(None),
+        }
+    }
+
+    #[test_log::test(tokio::test)]
+    async fn test_connect() {
+        static CONNECTION_COUNT: AtomicUsize = AtomicUsize::new(0);
+        static CONNECTED: AtomicBool = AtomicBool::new(true);
+        struct Client;
+
+        #[async_trait]
+        impl SimpleClient for Client {
+            async fn connect() -> Result<Self> {
+                CONNECTION_COUNT.fetch_add(1, Ordering::SeqCst);
+                Ok(Client)
+            }
+
+            fn is_connected(&self) -> bool {
+                CONNECTED.load(Ordering::SeqCst)
+            }
+
+            async fn request(&self, _method: &str) -> Result<String> {
+                Ok("null".to_string())
+            }
+        }
+
+        let fed = federation_member::<Client>();
+        assert_eq!(
+            CONNECTION_COUNT.load(Ordering::SeqCst),
+            0,
+            "should not connect before first request"
+        );
+
+        fed.request::<()>("", &[]).await.unwrap();
+        assert_eq!(
+            CONNECTION_COUNT.load(Ordering::SeqCst),
+            1,
+            "should connect once after first request"
+        );
+
+        fed.request::<()>("", &[]).await.unwrap();
+        assert_eq!(
+            CONNECTION_COUNT.load(Ordering::SeqCst),
+            1,
+            "should not connect again before disconnect"
+        );
+
+        // disconnect
+        CONNECTED.store(false, Ordering::SeqCst);
+
+        fed.request::<()>("", &[]).await.unwrap();
+        assert_eq!(
+            CONNECTION_COUNT.load(Ordering::SeqCst),
+            2,
+            "should connect again after disconnect"
+        );
+    }
+
+    #[test_log::test(tokio::test)]
+    async fn concurrent_requests() {
+        static CONNECTION_COUNT: AtomicUsize = AtomicUsize::new(0);
+        lazy_static! {
+            static ref FAIL: Mutex<HashSet<usize>> = Mutex::new(HashSet::new());
+        }
```

Comment:
Similarly, this could probably be an `Arc<Mutex<…>>` inside `Client` avoiding the need for `lazy_static`.